Update of Data Validation and binding in general

grokys opened the pull request:
I originally set out to implement the system in #529 but it turned out to be a lot more difficult than I expected :) In the end this has ended up being quite a major refactoring of our binding infrastructure, but hopefully for the better.

The changes on the whole follow the outline in #529, but I added a few other features too:
- Added an error adornment for TextBox
- Added support for validating Data Annotations
- Improved the way errors are logged (this may still need some work though)
- Data validation is currently only supported for direct properties
- Data validation is now automatic for properties that support it:
# How it will look to a control implementer

Data validation will be enabled per-property. In addition, for the moment data validation will only
be available on direct properties. To enable data validation on a property, set the
`EnableDataValidation` flag on property registration:

``` csharp
public static readonly DirectProperty<TextBox, string> TextProperty =
    AvaloniaProperty.RegisterDirect<TextBox, string>(
        nameof(Text),
        o => o.Text,
        (o, v) => o.Text = v,
        enableDataValidation: true);

private string _text;

public string Text
{
    get { return _text; }
    set { SetAndRaise(TextProperty, ref _text, value); }
}
```

Once data validation is enabled on a property, any binding to that property will have data
validation enabled, however nothing is done with the validation information by default: as mentioned
before, it is "pay to play" so nothing gets stored unless the control decides to. Instead, the
framework calls the `UpdateDataValidation` method each time a validated binding value is received.
In the example below, we use that information to set the `:error` pseudoclass on the control
depending on the validation state:

``` csharp

protected override void UpdateDataValidation(AvaloniaProperty property, BindingNotification status)
{
    // Set the ":error" pseudoclass if the error status is anything but "none".
    if (property == TextProperty)
    {
        ((IPseudoClasses)Classes).Set(":error", status.ErrorType != BindingErrorType.None);
    }
}
```

Commit:
Merge branch 'master' into indexer-bindings

Commit:
Added BindingNotification.

This will replace BindingError and IValidationStatus for bindings that want to notify binding/validation errors. Still WIP.

grokys committed:
Reimplemented data validation using BindingNotifications.

grokys committed:
Use NotifyingBase in unit tests.

And rename NotifyingBase.SubscriptionCount ->
PropertyChangedSubscriptionCount as we also track ErrorsChanged
subscription count.

grokys committed:
Added IndeiBase

And use it as base class for test data that implements
INotifyDataErrorInfo. Removed IndeiValidatorTests as its been
superceeded by IndeiValidationPluginTests.

grokys committed:
Started refactor of ExpressionObserver.

Tying to be more "rx", but also will allow us to move forward on BindingNotification changes.

grokys committed:
Add concept of "value plugins".

These are used to extensibly handle special values like `Task` and
`IObservable<>`. Previously this was baked into the expression observer
architecture with a TODO comment saying that it needs to be extensible.

grokys committed:
Subscribe to ExpressionSubject before setting value.

The inner `ExpressionObserver` must be subscribed to as setting the
value requires the expression to be evaluated.

grokys committed:
Fix MemberSelector.

grokys committed:
Updated some docs.

And simplified a bit of failing test code.

grokys committed:
Fix failing tests, add other tests.

Also use logical parent's DataContext as base for DataContext bindings
instead of visual parent's. Was previously in error.

grokys committed:
Removed vestiges of old data validation.

grokys committed:
Surface BindingNotifications in AvaloniaObject.

grokys committed:
Handle BindingNotifications in ExpressionSubject.

grokys committed:
Pass EnableDataValidation to ExpressionObserver.

grokys committed:
Fix casting of BindingNotification value.

grokys committed:
Correctly convert BindingNotifications.

In ExpressionSubject.

grokys committed:
Log meaningful message for AggregateException.

grokys committed:
Test that BindingNotificationReceived is called.

For direct properties.

grokys committed:
Fixed typo.

grokys committed:
Only send BindingNotifications on error.

Assume that a normal value when received by a property with data
validation enabled means no error.

grokys committed:
FIx failing test.

Introduced in a refactoring a few commits ago where I saw code that
looked wrong. It wasn't wrong, it's just the mechanism is a little
unintuitive. Added a comment explaining the reasoning.

grokys committed:
Add missing file to .csproj.

grokys committed:
Added more data validation tests, some failing.

grokys committed:
Added missing file in .csproj.

grokys committed:
Handle BindingNotifications in SetValue.

As one-time bindings don't set up a binding as such: they just call
`SetValue`.

grokys committed:
Remove data validation for non-direct properties.

We're going to say that for the moment only direct properties handle
data validation. This gets around a few thorny issues with data
validation on styled properties.

grokys committed:
Update BindingTest with changes.

Data validation still doesn't work in BindingTest though.

grokys committed:
Started to implement data validation on TextBox.

Not working still - added a failing test to demonstrate it.

grokys committed:
Make sure new value is returned.

When a property value is set, if a new value isn't notified via INPC,
then send the new value after the set succeeds.

grokys committed:
Revert "Only send BindingNotifications on error."

This reverts commit 57e646583fcd540a9b2795ee5e07700e241ea4f7.

For TextBox there can be 2 bindings to Text: one which is the binding to the
view model which should have data validation enabled, and another binding to
the TextPresenter in the template which should not have data validation
enabled, or it will override the view model data validation. For this we need
to be able to distinguish between the two and so bindings with data validation
enabled need to always send BindingNotifications.

Conflicts:
	src/Avalonia.Base/AvaloniaObject.cs
	tests/Avalonia.Markup.Xaml.UnitTests/Data/BindingTests_Validation.cs

grokys committed:
Update doc comment.

grokys committed:
`status` will no longer be null.

grokys committed:
Always call UpdateDataValidation on notification.

This speeds up the common path of no data validation: this way we don't
need to get the metadata unless a binding notification is received. This
means that if a binding with data validation is assigned to a property
that isn't interested then UpdateDataValidation will be called, but the
control should just ignore it.

grokys committed:
Only enable data validation for LocalValue bindings.

grokys committed:
Use `:error` pseudoclass in TextBox.

grokys committed:
Added INDEI validation to BindingTest.

grokys committed:
Moved some leak tests to LeakTests.

So they can be debugged in dotMemory.

grokys committed:
Convert to WeakReference in ExpressionObserver.

Doing a `Publish().Refcount()` caches the latest value so it can be sent
to subsequent subscribers, and this causes a leak. Convert the value
to/from `WeakReference` for that part.

grokys committed:
Use WeakReference in BindingNotification.

So that it doesn't keep objects alive when cached by
`.Publish().RefCount()` in `ExpressionObserver`. Added a leak test to
test that.

grokys committed:
Don't modify text due to binding when typing.

If a TextBox is e.g. bound to an int and the user types "02" then the
TextBox text should be "02" - not the value that comes back from the
binding which will be "2".

grokys committed:
Started adding TextBox error adornments.

grokys committed:
Fix the TextBox error ToolTip.

Create a new popup window for each ToolTip as sharing one was causing
problems when ToolTip was included in a control template.

grokys committed:
Fix TextBox/error icon scrolling.

grokys committed:
Fix OneTime bindings.

OneTime bindings were failing in BindingTest because the initial binding
error was being counted as the single value to transfer. Don't do this
with OneTime bindings - only transfer valid values.

grokys committed:
Move Avalonia.Markup.UnitTests to .net 4.5.

grokys committed:
Add string representation of BindingNotification.

For easier debugging.

grokys committed:
Added Data Annotations validation.

grokys committed:
Unpack AggregateException in TextBox.

grokys committed:
Fix logic in setting templated parent in Popup.

The brackets were in the wrong place!

grokys committed:
Don't clear TemplatedParent on removal from logical tree.

This meant that if a control with a template was removed, when it gets
re-added all of the template children's bindings are broken. The logic
to clear the TemplatedParent when applying a new template is in
TemplatedControl.ApplyTemplate - this is where it should be done.

grokys committed:
Always use an IList (array) for DataValidationErrors.

It's dangerous to pass a non-collection IEnumerable to an ItemsControl
and will be made illegal at some point.

grokys committed:
Don't clear ItemContainerGenerator on template change.

Can't remember why this was there but it was wrong anway - it wasn't
actually removing the items from the presenter panel. Doesn't seem to
break any unit tests so hope for the best...

grokys committed:
Improve broken binding diagnostic messages.

grokys committed:
Filter out broken binding errors in TextBox.

grokys committed:
Try to improve logged binding errors.

- Don't log an error when the target for the root ExpressionNode is
null. This is usually because the `DataContext` hasn't been set up yet
and it spewed a load of useless error messages.
- Add a Description field to `ExpressionObserver` that can be used in
the case of e.g. #control bindings to record the whole expression (with
the "#control" part) rather than just the part tracked by the
`ExpressionObserver`.

grokys committed:
Improve numerical binding error messages.

grokys committed:
Renamed ExpressionSubject -> BindingExpression.

`ExpressionSubject` wasn't a good name as the important thing it isn't
convert an `ExpressionObserver` into an `ISubject`. The important thing
it does is insert an `IValueConverter` into the pipeline.

Commit:
Merge branch 'master' into binding-updates2

Conflicts:
	src/Avalonia.Controls/TextBox.cs
	tests/Avalonia.Controls.UnitTests/TextBoxTests.cs

Commit:
Merge branch 'relative-path-fixes' into binding-updates2

Commit:
Use a path for the TextBox error icon.

grokys committed:
Merge branch 'relative-path-fixes' into binding-updates2

jkoritzinsky commented on src/Markup/Avalonia.Markup/Data/ExpressionNode.cs:
```diff
@@ -2,124 +2,188 @@
+            _observer.OnNext(value);
+        }
+
+        private IDisposable StartListeningCore()
```
I think that the names of `StartListeningCore` and `StartListening` should be swapped. Method names ending with `Core` should be the subclass implementation.

jkoritzinsky commented:
Other than the naming of `StartListeningCore` and `StartListening` this looks good to me!

There's some other features that I'd like to add to it, but considering the massive size of this PR those should wait until after this is merged.

grokys committed:
Swapped names in ExpressionNode.

Swapped names of StartListening and StartListeningCore - method names
ending with Core should be the subclass implementation.

grokys committed:
Merge branch 'master' into binding-updates2

grokys committed:
Add ncrunchprojects for new projects.

They need to be ignored by NCrunch.
